src/plugins/intel_gpu/src/kernel_selector/kernels/concatenation/concatenation_kernel_base.cpp

```diff
@@ -47,12 +47,38 @@ bool ConcatenationKernelBase::Validate(const Params& p, const optional_params&)
 }
 
 JitConstants ConcatenationKernelBase::GetJitConstants(const concatenation_params& params) const {
-    JitConstants jit = MakeBaseParamsJitConstants(params);
+    auto& inputs = params.original_input_layouts;
+    bool is_dynamic = std::any_of(inputs.begin(), inputs.end(), [](const DataTensor& t) { return t.is_dynamic(); }) ||
+                      std::any_of(params.outputs.begin(), params.outputs.end(), [](const DataTensor& t) { return t.is_dynamic(); });
+    JitConstants jit = MakeBaseParamsJitConstants(params, !is_dynamic);
 
     jit.AddConstants({
         MakeJitConstant("CONCAT_" + toString(params.axis), 1),
     });
 
+    if (is_dynamic) {
+        // shape info is supposed to contain shapes for
+        // in0, in1, ..., inN, out0
+        // So each dynamic kernel requires some custom offsets for shape_info access
+        size_t in_offset = 0;
+        for (size_t i = 0; i < params.kernel_split_id; i++) {
+            if (params.original_input_layouts[i].is_dynamic() || params.original_input_layouts[i].LogicalSize() == 0)
+                in_offset++;
+        }
+        size_t out_offset = 0;
+        for (size_t i = 0; i < params.original_input_layouts.size(); i++) {
+            if (params.original_input_layouts[i].is_dynamic() || params.original_input_layouts[i].LogicalSize() == 0)
+                out_offset++;
+        }
+
+        jit.AddConstant(MakeJitConstant("INPUT0", params.inputs[0], in_offset));
+        jit.AddConstant(MakeJitConstant("OUTPUT", params.outputs[0], out_offset));
+
+        jit.AddConstant(MakeJitConstant("IS_DYNAMIC", 1));
+        jit.AddConstant(MakeJitConstant("OPTIONAL_SHAPE_INFO_ARG", "__global const int* shape_info,"));
```

Comment:
Isnt this generated by common kernel_base.cpp?

Reply:
Common tensor definitions are not generated for that case:
```cpp
JitConstants jit = MakeBaseParamsJitConstants(params, !is_dynamic); // !is_dynamic disables it
```